tools/pin_github_action.py

```diff
@@ -0,0 +1,88 @@
+from __future__ import annotations
+
+import os
+
+GITHUB_TOKEN = os.environ.get("GITHUB_TOKEN")
+if not GITHUB_TOKEN:
+    raise SystemExit("GITHUB_TOKEN not set.")
+
+import argparse
+import json
+import re
+import urllib.request
+from functools import lru_cache
+from typing import Sequence
+from urllib.error import HTTPError
+
+
+@lru_cache(maxsize=None)
+def get_sha(repo: str, ref: str) -> str:
+    if len(ref) == 40:
+        try:
+            int(ref, 16)
+            return ref
+        except ValueError:
+            pass
+
+    try:
+        resp = urllib.request.urlopen(
+            urllib.request.Request(
+                f"https://api.github.com/repos/{repo}/commits/{ref}",
+                method="GET",
+                headers={
+                    "Accept": "application/vnd.github+json",
+                    "Authorization": f"token {GITHUB_TOKEN}",
+                    # A user agent is required. Lol.
+                    "User-Agent": "python-requests/2.26.0",
+                },
+            )
+        )
+    except HTTPError as e:
+        print(f"Status {e.code} while resolving {ref} for {repo}.")
+        if e.code == 403:
+            print("You most likely didn't authorize your token for SAML to the getsentry org.")
+        return ref
+    if resp.status != 200:
+        print(f"Status {resp.status} while resolving {ref} for {repo}.")
+        return ref
+
+    data = json.load(resp)
+    return data["sha"]  # type: ignore
```

Comment:
rather than `# type: ignore` this should technically be a `cast` -- or you can do `data: dict[str, str] = json.load(...)` and avoid the cast or ignore

Reply:
oh! yes, i wanted to do something like that but couldn't find out how